src/libcrun/ebpf.c

```diff
@@ -293,15 +293,90 @@ bpf_program_complete_dev (struct bpf_program *program, libcrun_error_t *err arg_
   return program;
 }
 
+static int
+ebpf_attach_program (int fd, int dirfd, libcrun_error_t *err)
+{
+  const int MAX_ATTEMPTS = 20;
+  int attempt;
+
+  for (attempt = 0;; attempt++)
+    {
+      cleanup_close int replacefd = -1;
+      union bpf_attr attr;
+      uint32_t progs[2];
+      int ret;
+
+      memset (&attr, 0, sizeof (attr));
+      attr.query.target_fd = dirfd;
+      attr.query.attach_type = BPF_CGROUP_DEVICE;
+      attr.query.prog_cnt = sizeof (progs) / sizeof (progs[0]);
+      attr.query.prog_ids = (uint64_t) &progs;
+
+      ret = bpf (BPF_PROG_QUERY, &attr, sizeof (attr));
+      if (UNLIKELY (ret < 0))
+        return crun_make_error (err, errno, "bpf query");
+
+      if (attr.query.prog_cnt > 1)
+        return crun_make_error (err, 0, "invalid device cgroup state, more than one program installed");
+
+      if (attr.query.prog_cnt == 1)
+        {
+#ifdef BPF_F_REPLACE
+          memset (&attr, 0, sizeof (attr));
+          attr.prog_id = progs[0];
```

Comment:
This logic (replacing whatever program is attached) and the corresponding limitations (not being able to replace a program if there is more than one) seems a bit fishy. I think a much better solution would be to either pin the program, or to store the id and load-time into some state file (which I assume `crun` has) and then to use that to fetch the original program.

To be fair, it's unlikely anybody else will be attaching `BPF_CGROUP_DEVICE` programs to a container but this seems a bit fishy to me.